File: integration_tests/assumed_type_03.f90
program assumed_type_03
  ! MRE for caffeine co_broadcast double-free with LFortran.
  !
  ! Derived types with fixed-length character components must store the
  ! character data inline (like gfortran does) so that raw byte
  ! operations through type(*) / CFI descriptors work correctly.
  ! If the character is stored as a heap pointer instead, memcpy-style
  ! operations (e.g. caf_co_broadcast) copy the pointer, not the data,
  ! leading to double-free and wrong results across images.
  implicit none

  type :: pair_t
    integer :: id
    character(len=5) :: label
  end type

  interface
    subroutine check_dt_char_inline(a, ok) bind(C)
      type(*), intent(in) :: a(..)
      integer, intent(out) :: ok
    end subroutine

    subroutine copy_via_cfi(src, dst) bind(C)
      type(*), intent(in)    :: src(..)
      type(*), intent(inout) :: dst(..)
    end subroutine
  end interface

  type(pair_t) :: src, dst
  integer :: ok

  src = pair_t(42, "hello")

  call check_dt_char_inline(src, ok)
  if (ok /= 1) error stop "character data not stored inline in derived type"

  dst = pair_t(0, "xxxxx")
  call copy_via_cfi(src, dst)
  if (dst%id /= 42) error stop "integer component wrong after raw copy"
  if (dst%label /= "hello") error stop "character component wrong after raw copy"

  print *, "ok"
end program

File: integration_tests/assumed_type_03.c
#include <ISO_Fortran_binding.h>
#include <string.h>
#include <stdint.h>

void check_dt_char_inline(CFI_cdesc_t *desc, int32_t *ok) {
    /*
     * Verify that the 5-byte string "hello" appears physically inside the
     * raw struct bytes described by the CFI descriptor.  If the Fortran
     * compiler stores character(len=5) inline in the derived type (correct
     * per Fortran standard for non-pointer, non-allocatable components),
     * the bytes are part of the struct and we find them.  If the compiler
     * stores a heap pointer instead, the struct contains an address, not
     * the character data, and this search fails.
     */
    const char *base = (const char *)desc->base_addr;
    size_t len = desc->elem_len;
    *ok = 0;
    for (size_t i = 0; i + 5 <= len; i++) {
        if (memcmp(base + i, "hello", 5) == 0) {
            *ok = 1;
            return;
        }
    }
}

void copy_via_cfi(CFI_cdesc_t *src, CFI_cdesc_t *dst) {
    /*
     * Raw byte copy — exactly what GASNet's gex_Coll_BroadcastNB does
     * inside caffeine's caf_co_broadcast.
     */
    memcpy(dst->base_addr, src->base_addr, src->elem_len);
}
